Scraper & processor

ashersnavely opened the pull request:
Just overall process/syntax review sometime tomorrow.

Commit:
Retrying failed push from last time

Commit:
Fixed references, should be ready for debug

Commit:
Adding my stuff back in.
Kind of one huge push here since my work has been pretty isolated so far.

Commit:
Fixed bugs in current stage, ready for deeper testing and finishing off tomorrow.

Commit:
Cleaned up a bit before calling it quits for now.

Commit:
Release threads for client balancing are now tested and working.

johnsimmons2 reviewed (commented):
There is a lot of redundant classes here, StackOverflow shows up more than once as does ReleaseHeap. Did you separate them out into other files but not delete their original occurrences?

Commit:
Worked on delay between requests set to spread them out between the given time.

johnsimmons2 commented:
> There is a lot of redundant classes here, StackOverflow shows up more than once as does ReleaseHeap. Did you separate them out into other files but not delete their original occurrences?

I see /scraping and /scraper, is there a difference? Do these classes need 2 definitions

johnsimmons2 commented on stackoversight/scraper/release_heap.py:
```diff
@@ -0,0 +1,48 @@
+from queue import Queue
+
+
+class ReleaseHeap(object):
```
I notice there is ReleaseHeap(object) and ReleaseHeap(queue) (I do not know if doing (object) is standard notation or something, but it is redundant and not needed for Python 3) do both classes need to exist? If so consider changing their names to show explicit differences

johnsimmons2 replied:
This is resolved with latest push

johnsimmons2 commented on stackoversight/scraping/site/Site.py:
```diff
@@ -0,0 +1,180 @@
+        return content
+
+
+class StackOverflow(Site):
```
2nd occurrence of StackOverflow(Site) class definition

Commit:
Improved delay method.

ashersnavely commented:
Accidentally left in old files, thanks for catching that.

They should be removed now!

Commit:
Removed accidentally left in files!

johnsimmons2 reviewed (commented):
2 questions/comments but otherwise it is good to merge

ashersnavely commented on stackoversight/scraper/scraper.py:
```diff
@@ -0,0 +1,35 @@
+    for text in site.get_text(soup):
+        print(text)
+
+print(site.balancer.heap)
```
Just for testing!

ashersnavely commented on stackoversight/scraper/release_heap.py:
```diff
@@ -0,0 +1,48 @@
+        self.release_queue = Queue()
+
+        # quick access to pair by it's value
+        self.hash_map = dict([(pair[1], pair) for pair in self.heap])
```
Yes, there is actually a class called "mutabletuple" and all i did was replace the non-mutable tuples with those.
Made sure to leave a comment about the type requirement before the ReleaseHeap init method.

johnsimmons2 replied:
I saw that elsewhere, but why is there no mutabletuple here? @ashersnavely just curious

Commit:
Large change here, making requests use OAuth2. A lot needs to be cleaned and looked over and debugged.

Commit:
Working on using credentials, having issues fetching tokens.

johnsimmons2 commented:
@ashersnavely is this closer to being able to pull and close?

Commit:
Finally got it working with client key.

Realized oauth wouldnt work out a while into it because it is really for user clients, the mix up was only because im completely new to these thing, and so had to change gears back to what I was originally doing.

Besides replacing the oauth setup with client keys i also implemented a check on key initialization to get the remaing requests under the quota limit which is really good for unexpected stopping and starting of the scraping process

Commit:
Continued implementation of client keys and looking into expanding the functionality of the stackoverflow site class

Commit:
Working on changing things up a bit to conform to StackExhange's API more, goal is for this to be able to be run indefinitely after all so needs to listen to things like their back_off responses and such

Commit:
Now using the stackexchange api correctly instead of scraping stackoverflow directly

Commit:
Done for tonight, need to fix parsing the response now that were using stackexchange instead of directly scraping stackoverflow.

Next big thing to do is to use the info returned in the response such as back_off has_more etc...

Commit:
Successfully using the stackexhange api to get links and then directly scraping those links, needs to be optimized and automated more

Commit:
Cleaned up

Commit:
Handles backoff response now

Commit:
Quick implementation for Commander Black, needed to start pulling data asap

Commit:
Quick implementation done, ready to start scraping. Going to adjust the timing to make a little faster than wed normally do
